### PR TITLE
feat(storage): support sstable store with local object store

### PR DESCRIPTION
**Pull request by wenym1**
## What's changed and what's your intention?

In this PR, we added a local object store to sstable store. When we operate on sstable store, we use the highest bit in sstable id to indicate whether we use the local or remote object store.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
part of #2384

### CODE REVIEW

**Comment by hzxa21** [comment]
What is the relationship between this PR and #2389?

**Comment by wenym1** [comment]
> What is the relationship between this PR and #2389?

#2389 simply added a new disk object store. This PR add a local object store in sstable store, and such local object store is not necessarily disk object store.

File: src/storage/src/hummock/sstable_store.rs [hunk 1]
```diff
@@ -108,6 +112,42 @@ impl SstableStore {
         Ok(len)
     }
 
+    pub async fn put_meta(&self, sst: &Sstable) -> HummockResult<()> {
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable_store.rs:115]: nits: can we make `put_meta`, `put_sst_data`, `delete_sst_data` private?

**Review by hzxa21 — COMMENTED** [review]
Generally LGTM.

File: src/storage/src/hummock/compactor.rs [hunk 2]
```diff
@@ -145,17 +145,41 @@ impl Compactor {
...
             let iter = {
                 let iters = buffers.iter().map(|m| {
                     Box::new(m.clone().into_forward_iter()) as BoxedForwardHummockIterator
                 });
                 MergeIterator::new(iters, stats.clone())
             };
             let vnode2unit = vnode2unit.clone();
+            let compactor = compactor.clone();
             compaction_futures.push(tokio::spawn(async move {
+                let builder = GroupedSstableBuilder::new(
```
Comment (Little-Wallace) [inline src/storage/src/hummock/compactor.rs:157]: Why add `GroupedSstableBuilder` in this PR? It seems not belong to this PR

File: src/storage/src/hummock/sstable_store.rs [hunk 3]
```diff
@@ -57,10 +68,12 @@ impl SstableStore {
         ));
         Self {
             path,
-            store,
+            remote_store,
+            local_store,
             block_cache: BlockCache::new(block_cache_capacity),
             meta_cache,
             stats,
+            next_local_sst_id: AtomicU64::new(0),
```
Comment (Little-Wallace) [inline src/storage/src/hummock/sstable_store.rs:76]: do not maintain `next_local_sst_id` in `SstableStore`.
It belongs to `Version`. Manage it in `LocalVersionManager` or other place

File: src/storage/src/store_impl.rs [hunk 4]
```diff
@@ -92,11 +92,14 @@ impl StateStoreImpl {
     ) -> StorageResult<Self> {
         let store = match s {
             hummock if hummock.starts_with("hummock+") => {
-                let object_store = Arc::new(
+                let remote_object_store = Arc::new(
```
Comment (Little-Wallace) [inline src/storage/src/store_impl.rs:95]: StateStore does not need to concern about `remote_object_store` and `local_object_store`. If you want to flush some data to local object store, it's work of `WriteBufferUploader`

File: src/storage/src/hummock/compactor.rs (hunk 2 above)
Comment (wenym1) [reply src/storage/src/hummock/compactor.rs:157]: The latest main branch has already used this `GroupedSstableBuilder `. It's simply a refactor.

Comment (Little-Wallace) [reply src/storage/src/hummock/compactor.rs:157]: ok

**Comment by Little-Wallace** [comment]
@hzxa21  Since in some case of our product, we may not need two object store, in other word, `spill-to-disk` may not make sense for all users, I can not agree to always keep two kind of object store in `SstableStore`. In most condition, we only need one remote storage for all.

File: src/storage/src/hummock/sstable_store.rs (hunk 3 above)
Comment (wenym1) [reply src/storage/src/hummock/sstable_store.rs]: Have moved it to `SharedBufferUploader`

File: src/storage/src/hummock/sstable/builder.rs [hunk 5]
```diff
@@ -37,12 +38,25 @@ pub struct SSTableBuilderOptions {
...
     /// Restart point interval.
     pub restart_interval: usize,
-    /// False prsitive probability of bloom filter.
+    /// False positive probability of bloom filter.
     pub bloom_false_positive: f64,
     /// Compression algorithm.
     pub compression_algorithm: CompressionAlgorithm,
 }
 
+impl SSTableBuilderOptions {
```
Comment (Little-Wallace) [inline src/storage/src/hummock/sstable/builder.rs:47]: Impl `From`?

**Review by Little-Wallace — APPROVED** [review]
LGTM
